### PR TITLE
Add extras_require for various worker types

### LINKED ISSUE #1717: Use extras_require in setup for optional dependencies
Gunicorn [may require extra dependencies](http://docs.gunicorn.org/en/stable/install.html#async-workers) to be installed, but they seem to be optional. It looks like it should be possible to install these using [setuptools's extra](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies).

It looks like Gevent or Eventlet are 2 options, but the [worker option](http://docs.gunicorn.org/en/stable/run.html#commonly-used-arguments) seems to indicate there are more.

Would it make sense to add this? It has the advantage of documenting indirect dependencies.

### PR DESCRIPTION
**Pull request by browniebroke**
Fixes #1717

- [x] Add [`extras_require`](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) to `setup.py`
- [x] Update documentation

### CODE REVIEW

File: docs/source/settings.rst [hunk 1]
```diff
@@ -119,10 +119,12 @@ you might want to choose one of the other worker classes.
 A string referring to one of the following bundled classes:
 
 * ``sync``
-* ``eventlet`` - Requires eventlet >= 0.9.7
-* ``gevent``   - Requires gevent >= 0.13
-* ``tornado``  - Requires tornado >= 0.2
-* ``gthread``  - Python 2 requires the futures package to be installed
+* ``eventlet`` - Requires eventlet >= 0.9.7 (or install using
```
Comment (berkerpeksag) [inline docs/source/settings.rst]: Could you please make these changes in `gunicorn/config.py` and then run `make html` in `docs` directory? (and please  don't include unrelated changes to `docs/source/settings.rst` after you ran `make html`)

File: setup.py [hunk 2]
```diff
@@ -72,6 +72,16 @@ def run_tests(self):
         errno = pytest.main(self.test_args)
         sys.exit(errno)
 
+
+extra_require = {
+    'gevent':  ["gevent>=0.13"],
+    'eventlet': ["eventlet>=0.9.7"],
+    'tornado': ["tornado>=0.2"],
+    'gthread': []
```
Comment (berkerpeksag) [inline setup.py]: Style nit: Please add trailing comma.

File: setup.py [hunk 3]
```diff
@@ -72,6 +72,16 @@ def run_tests(self):
         errno = pytest.main(self.test_args)
         sys.exit(errno)
 
+
+extra_require = {
+    'gevent':  ["gevent>=0.13"],
```
Comment (berkerpeksag) [inline setup.py]: Style nit: We usually try to stick with the most used quoting style in a file. For `setup.py`, it looks like we went with single quotes. It would be nice use it in new code.

File: setup.py [hunk 4]
```diff
@@ -98,5 +108,6 @@ def run_tests(self):
 
     [paste.server_runner]
     main=gunicorn.app.pasterapp:paste_server
-    """
+    """,
+    extras_require=extra_require
```
Comment (berkerpeksag) [inline setup.py]: Style nit: Please add trailing comma.

File: docs/source/install.rst [hunk 5]
```diff
@@ -33,12 +33,12 @@ want to consider one of the alternate worker types.
 
 ::
 
-    $ pip install greenlet  # Required for both
-    $ pip install eventlet  # For eventlet workers
-    $ pip install gevent    # For gevent workers
+    $ pip install gunicorn[eventlet]  # For eventlet workers
+    $ pip install gunicorn[gevent]    # For gevent workers
```
Comment (benoitc) [inline docs/source/install.rst]: i’m not sure about this changes. some people are installing gunicorn with both workers support and not always at the same time. I would suggest to keep it like this and maybe add a new line.

File: docs/source/settings.rst (hunk 1 above)
Comment (benoitc) [reply docs/source/settings.rst]: indeed that should be done in settings.py in any case

Comment (browniebroke) [reply docs/source/settings.rst]: > please don't include unrelated changes to `docs/source/settings.rst` after you ran `make html`

Looking at the generated `settings.rst`, I'm not sure what you mean. Do you mean I should not commit it at all?

Comment (berkerpeksag) [reply docs/source/settings.rst]: After you ran `make html`, if you do `git diff` on `settings.rst` you will see changes like:

```diff
diff --git a/docs/source/settings.rst b/docs/source/settings.rst
index 4a8d027..738e18a 100644
--- a/docs/source/settings.rst
+++ b/docs/source/settings.rst
@@ -450,7 +450,7 @@ chdir
 ~~~~~
 
 * ``--chdir``
-* ``/Users/benoitc/work/gunicorn/py27/gunicorn/docs``
+* ``/home/berker/projects/gunicorn/docs``
 
 Chdir to specified directory before apps loading.
 
@@ -521,7 +521,7 @@ user
 ~~~~
 
 * ``-u USER, --user USER``
-* ``501``
+* ``1000``
 
 Switch worker processes to run as this user.
 
@@ -535,7 +535,7 @@ group
 ~~~~~
 
 * ``-g GROUP, --group GROUP``
-* ``20``
+* ``1000``
 
 Switch worker process to run as this group.
 
@@ -594,7 +594,7 @@ temporary directory.
 secure_scheme_headers
 ~~~~~~~~~~~~~~~~~~~~~
 
-* ``{'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}``
+* ``{'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on', 'X-FORWARDED-PROTOCOL': 'ssl'}``
 
 A dictionary containing headers and values that the front-end proxy
 uses to indicate HTTPS requests. These tell Gunicorn to set
@@ -763,13 +763,30 @@ The log config file to use.
 Gunicorn uses the standard Python logging module's Configuration
 file format.

```

I was talking about these unrelated changes in my previous comment. They are created by `docs/gunicorn_ext.py` but we usually discard them when we generate `settings.rst` for documentation changes.

Comment (browniebroke) [reply docs/source/settings.rst]: Ok, thanks for the clarification. Should be better now.

File: gunicorn/config.py [hunk 6]
```diff
@@ -605,15 +605,17 @@ class WorkerClass(Setting):
 
         The default class (``sync``) should handle most "normal" types of
         workloads. You'll want to read :doc:`design` for information on when
-        you might want to choose one of the other worker classes.
+        you might want to choose one of the other worker classes. Required
+        libraries may be installed using setuptools' extra_require feature.
```
Comment (berkerpeksag) [inline gunicorn/config.py]: extra_require -> \`\`extra_require\`\`

File: gunicorn/config.py [hunk 7]
```diff
@@ -605,15 +605,17 @@ class WorkerClass(Setting):
...
 
         A string referring to one of the following bundled classes:
 
         * ``sync``
-        * ``eventlet`` - Requires eventlet >= 0.9.7
-        * ``gevent``   - Requires gevent >= 0.13
-        * ``tornado``  - Requires tornado >= 0.2
+        * ``eventlet`` - Requires eventlet >= 0.9.7 (extra gunicorn[eventlet])
```
Comment (berkerpeksag) [inline gunicorn/config.py]: Maybe something like "(or install it via \`\`gunicorn[...]\`\`)" or "(or install it via \`\`pip install gunicorn[...]\`\`)" would make it clearer.

Comment (browniebroke) [reply gunicorn/config.py]: Thanks for the suggestion, I wasn't happy with my wording myself.

**Comment by berkerpeksag** [comment]
Thank you and congratulations for your first contribution to Gunicorn!
